torchx/test/torchx_test.exs

```diff
@@ -0,0 +1,144 @@
+  end
+
+  describe "binary tensor-tensor ops" do
+    for op <- @ops ++ @logical_ops,
+        type_a <- @types,
+        type_b <- @types,
+        not ((op in [:divide, :remainder, :atan2, :power] and (type_a == {:bf, 16} and type_b in @bf16_and_ints)) or
+               (type_b == {:bf, 16} and type_a in @bf16_and_ints)) do
+      test "#{op}(#{Nx.Type.to_string(type_a)}, #{Nx.Type.to_string(type_b)})" do
+        op = unquote(op)
+        type_a = unquote(type_a)
+        type_b = unquote(type_b)
+
+        test_binary_op(op, type_a, type_b)
+      end
+    end
+  end
+
+  describe "binary tensor-tensor bitwise ops" do
+    for op <- @bitwise_ops,
+        type_a <- @ints,
+        type_b <- @ints do
+      test "#{op}(#{Nx.Type.to_string(type_a)}, #{Nx.Type.to_string(type_b)})" do
+        op = unquote(op)
+        type_a = unquote(type_a)
+        type_b = unquote(type_b)
+
+        test_binary_op(op, type_a, type_b)
+      end
+    end
+  end
+
+  describe "unary ops" do
+    for op <- @unary_ops -- [:bitwise_not],
+        type <- @types do
+      test "#{op}(#{Nx.Type.to_string(type)})" do
+        test_unary_op(unquote(op), unquote(type))
+      end
+    end
+
+    for type <- @ints do
+      test "bitwise_not(#{Nx.Type.to_string(type)})" do
+        test_unary_op(:bitwise_not, unquote(type))
+      end
+    end
+  end
+
+  # Division with bfloat16 is a special case with PyTorch,
+  # because it upcasts bfloat16 args to float for numerical accuracy purposes.
+  # So, the result of division is different from what direct bf16 by bf16 division gives us.
+  # I.e. 1/5 == 0.19921875 in direct bf16 division and 0.2001953125 when dividing floats
+  # converting them to bf16 afterwards (PyTorch style).
+  describe "bfloat16" do
+    for type_a <- @bf16_and_ints,
+        type_b <- @bf16_and_ints,
+        type_a == {:bf, 16} or type_b == {:bf, 16} do
+      test "divide(#{Nx.Type.to_string(type_a)}, #{Nx.Type.to_string(type_b)})" do
+        type_a = unquote(type_a)
+        type_b = unquote(type_b)
+
+        a = tt([[1, 2], [3, 4]], type_a)
+        b = tt([[5, 6], [7, 8]], type_b)
+
+        c = Nx.divide(a, b)
+
+        assert Nx.backend_transfer(c) ==
+                 Nx.tensor([[0.2001953125, 0.333984375], [0.427734375, 0.5]],
+                   type: {:bf, 16}
+                 )
+      end
+    end
+
+    for type_a <- @bf16_and_ints,
+        type_b <- @bf16_and_ints,
+        type_a == {:bf, 16} or type_b == {:bf, 16} do
+      test "power(#{Nx.Type.to_string(type_a)}, #{Nx.Type.to_string(type_b)})" do
+        type_a = unquote(type_a)
+        type_b = unquote(type_b)
+
+        a = tt([[1, 2], [3, 4]], type_a)
+        b = tt([[5, 6], [7, 8]], type_b)
+
+        c = Nx.power(a, b)
+
+        assert Nx.backend_transfer(c) ==
+                 Nx.tensor([[1.0, 64.0], [2192.0, 65536.0]],
+                   type: {:bf, 16}
+                 )
+      end
+    end
+  end
+
+  describe "vectors" do
```

Comment:
What do you mean by vectors in this case? Outer only works with vectors for Pytorch?

Reply:
Yes, only with 1-D tensors. Should we reshape input tensors into vectors for it?

Reply:
Yes, only with 1-D tensors. Should we reshape input tensors into vectors for it?